Common/ExecutionModel/vtkThreadedImageAlgorithm.cxx

```diff
@@ -24,13 +23,104 @@
 #include "vtkObjectFactory.h"
 #include "vtkPointData.h"
 #include "vtkStreamingDemandDrivenPipeline.h"
+#include "vtkExtentTranslator.h"
+
+
+#define USE_SMP_DEFAULT_2D_BLOCK
+
+// the default 2d block is used to compare with tbb's own 2d splitting option
+#ifdef USE_SMP_DEFAULT_2D_BLOCK
+
+#include <tbb/parallel_for.h>
+#include <tbb/blocked_range2d.h>
+
+#else
+
+#include "vtkSMPTools.h"   
+
+#endif
+
+struct vtkImageThreadStruct
+{
+  vtkThreadedImageAlgorithm *Filter;
+  vtkInformation *Request;
+  vtkInformationVector **InputsInfo;
+  vtkInformationVector *OutputsInfo;
+  vtkImageData   ***Inputs;
+  vtkImageData   **Outputs;
+};
+
+#ifdef USE_SMP_DEFAULT_2D_BLOCK
+
+class vtkApplyTransformTBB
+{
+  vtkImageThreadStruct* Str;
+public:
+
+  vtkApplyTransformTBB(vtkImageThreadStruct* threadStruct)
```

Comment:
Rule 5 states that class definitions for helper classes go in the .cxx files (the "implementation" files), which is exactly what vtkSMPContourFilterManyPieces.cxx does.  So that file is following the rule, and currently your code is following the rule as well.